src/core/job.c

```diff
@@ -1173,7 +1178,32 @@ void job_add_to_run_queue(Job *j) {
                         log_warning_errno(r, "Failed to enable job run queue event source, ignoring: %m");
         }
 
-        LIST_PREPEND(run_queue, j->manager->run_queue, j);
+        /* Deterministic run-queue order:
+         * - mount and service units after any other units
+         * - units with higher CPU weight first
+         * - sort by id if the other criteria are the same */
+        state = manager_state(j->manager);
+        c = unit_get_cgroup_context(j->unit);
+        weight = c ? cgroup_context_cpu_weight(c, state) : CGROUP_WEIGHT_DEFAULT;
+        LIST_FOREACH(run_queue, tmp, j->manager->run_queue) {
+                if (!IN_SET(j->unit->type, UNIT_SERVICE, UNIT_MOUNT) &&
```

Comment:
This comparison should be split out into a separate function. It'll make the flow easier to follow.

We have four unit types that have exec sections: service, socket, mount, swap. Sockets are quick, but it seems all the other ones should be treated the same. Please add a comment (in the code) that explains the logic a bit.

Reply:
instead of doing these explicitly here, I'd prefer if we'd simply reorder the unit type enum by priority (and document that explicitly in the definition of the enum, so that people don't accidentaly reorder that again). If we did we could just do CMP(j->unit->type, tmp->unit->type) to figure out the prio between them

Reply:
Well, we're not sorting by type here. We have just two groups, but we can do a simple comparison instead of a IN_SET if the unit type is orders differently.

Reply:
This is completely different now and should be fixed.